app/gui2/src/components/ComponentBrowser.vue

```diff
@@ -194,6 +195,69 @@ function selectLastAfterRefresh() {
   })
 }
 
+// === Highlight matches ===
+
+const fullQualifiedNameQuery = computed(() => {
+  if (input.filter.value.qualifiedNamePattern != null) {
+    return input.filter.value.pattern != null
+      ? `${input.filter.value.qualifiedNamePattern}.${input.filter.value.pattern}`
+      : input.filter.value.qualifiedNamePattern
+  } else {
+    return input.filter.value.pattern
+  }
+})
+
+/** The first and last match are the parts of the string that are outside of the match.
+ * The middle matches come in groups of three, and contain respectively:
+ * - the unmatched prefix (must end with a `_`)
+ *   (an empty string if the entire qualified name segment was matched)
+ * - the matched text
+ * - the unmatched suffix (an empty string if the entire qualified name segment was matched)
+ * - the separator (`.` or `_`, or the empty string if this is the last segment) */
+const extractMatchesRegex = computed(() => {
+  if (fullQualifiedNameQuery.value == null) return undefined
+  let prefix = ''
+  let suffix = ''
+  for (const [, text, separator] of fullQualifiedNameQuery.value.matchAll(/(.+?)([._]|$)/g)) {
+    const segment =
+      separator === '_'
+        ? `()(${text})([^_.]*)(_)`
+        : `([^.]*_)?(${text})([^.]*)(${separator === '.' ? '\\.' : ''})`
+    prefix = '(?:' + prefix
+    suffix += segment + ')?'
+  }
+  return new RegExp('^(.*?)' + prefix + suffix + '(.*)$', 'i')
+})
+
+interface MatchHighlightSegment {
+  text: string
+  type: 'no-match' | 'match'
+}
+
+function* highlightMatches(name: QualifiedName): Generator<MatchHighlightSegment> {
+  const match =
+    extractMatchesRegex.value != null ? name.match(extractMatchesRegex.value) : undefined
+  if (match == undefined) {
+    yield { text: name, type: 'no-match' }
+    return
+  }
+  const prefix = match[1]
+  if (prefix) yield { text: prefix, type: 'no-match' }
+  const end = match.length - 4
+  for (let i = 2; i < end; i += 4) {
+    const prefix = match[i]
+    if (prefix) yield { text: prefix, type: 'no-match' }
+    const matched = match[i + 1]
+    if (matched) yield { text: matched, type: 'match' }
+    const suffix = match[i + 2]
+    if (suffix) yield { text: suffix, type: 'no-match' }
+    const separator = match[i + 3]
+    if (separator) yield { text: separator, type: 'no-match' }
```

Comment:
If user wrote the separator, I would treat it as matched.

Reply:
(this should be fixed in the new impl too)